src/ucp/rndv/proto_rndv.c

```diff
@@ -365,41 +348,147 @@ ucp_proto_rndv_ctrl_init(const ucp_proto_rndv_ctrl_init_params_t *params,
+    proto_config->priv           = priv;
+    proto_config->ep_cfg_index   = init_params->ep_cfg_index;
+    proto_config->rkey_cfg_index = init_params->rkey_cfg_index;
+    proto_config->select_param   = *select_param;
 }
 
-void ucp_proto_rndv_ctrl_probe(const ucp_proto_rndv_ctrl_init_params_t *params)
+void ucp_proto_rndv_ctrl_probe(const ucp_proto_rndv_ctrl_init_params_t *params,
+                               void *priv, size_t priv_size)
 {
-    ucp_proto_rndv_ctrl_priv_t priv;
-    ucp_proto_caps_t caps;
+    const ucp_proto_init_params_t *init_params = &params->super.super;
+    ucp_proto_rndv_ctrl_priv_t *rpriv          = priv;
+    const ucp_proto_select_init_protocols_t *remote_proto_init;
+    ucp_proto_select_param_t remote_select_param;
+    ucp_proto_select_elem_t *remote_proto_select;
+    ucp_proto_init_elem_t *remote_proto;
+    ucp_proto_caps_t result_caps, *caps;
+    ucp_proto_perf_range_t ctrl_perf;
     ucs_status_t status;
+    size_t min_length, max_length, cfg_thresh, cfg_priority;
+    void *remote_priv;
+    const char *proto_name;
+
+    ucs_assert(params->super.flags & UCP_PROTO_COMMON_INIT_FLAG_RESPONSE);
+    ucs_assert(!(params->super.flags & UCP_PROTO_COMMON_INIT_FLAG_SINGLE_FRAG));
 
-    status = ucp_proto_rndv_ctrl_init(params, &caps, &priv);
+    if (!ucp_proto_common_init_check_err_handling(&params->super)) {
+        return;
+    }
+
+    /* Initialize 'rpriv' structure */
+    ucp_proto_rndv_ctrl_init_priv(params, rpriv, params->lane);
+
+    /* Init remote proto */
+    remote_select_param = ucp_proto_rndv_remote_select_param_init(params);
+    status              = ucp_proto_rndv_ctrl_select_remote_proto(
+            params, &remote_select_param, rpriv->md_map, &remote_proto_select);
     if (status != UCS_OK) {
         return;
     }
+    remote_proto_init = &remote_proto_select->proto_init;
+
+    /* Add variants for each remote proto */
+    ucs_trace("add variants for %s: min_length=%zu max_length=%zu",
+              ucp_proto_id_field(init_params->proto_id, name),
+              params->super.min_length, params->super.max_length);
+    ucs_log_indent(+1);
+    ucs_array_for_each(remote_proto, &remote_proto_init->protocols) {
+        if (ucp_proto_id_field(remote_proto->proto_id, flags) &
+            UCP_PROTO_FLAG_INVALID) {
+            continue;
+        }
+
+        ucs_assertv(remote_proto->caps.num_ranges > 0,
+                    "remote_proto->caps.num_ranges=%u",
+                    remote_proto->caps.num_ranges);
+
+        proto_name = ucp_proto_id_field(remote_proto->proto_id, name);
+        caps       = &remote_proto->caps;
+        max_length = caps->ranges[caps->num_ranges - 1].max_length;
+        min_length = caps->min_length;
+        ucs_trace("%s proto variant: min_length=%zu max_length=%zu "
+                  "cfg_thresh=%zu cfg_priority=%u", proto_name, 
+                  min_length, max_length, remote_proto->cfg_thresh,
+                  remote_proto->cfg_priority);
```

Comment:
maybe move it below after checking `max_len > min_len`

Reply:
but you do not print `params->super.max_length` and `params->super.min_length`, so it maybe hard to see why some proto was disabled. Maybe add separate traces for the normal case and when there are not available ranges

Reply:
I print them on line 418

Reply:
ok